### ISSUE TITLE
Migrate from Heroku 

### ISSUE THREAD

**Issue by kerolloz** [icpc-scu-community/board-api#8]
Since Heroku is cancelling the free plan. We will have to migrate to another platform. 
The platform needs to support Redis and cron jobs (which is required for the scrapper).

Possible solutions:

- [Render.com](https://render.com)
  - Pros: Free. Easy to use. Supports Redis (includes a free tier). HTTPS for custom domains (unlike Heroku).
  - Cons: No free Cron jobs. Sleeps after 15 minutes of inactivity (same as Heroku) 

We can use [Deta.sh](https://deta.sh/) to overcome the shortages of Render. Deta.sh has free cron jobs. I am not really sure if it's suitable for our use case in the scrapper. We can also use Deta cron jobs to send requests to the API deployed on Render to force it not to sleep. Unfortunately, we can't use Deta.sh as our main backend because it [doesn't support](https://docs.deta.sh/docs/micros/about/#important-notes) connecting to MongoDB.
> P.S. I have personally tried Render & Deta

Let me know your thoughts @amrsalama @Waelahmed99 @7oSkaaa.

**Comment by Waelahmed99** [comment]
I can investigate other options over this week and give you a feedback list as you did with pros and cons, but since you personally used Render.com, I'm keen to this option.

It will be a painful job if we decide to migrate to [Deta base](https://docs.deta.sh/docs/base/about) right? (Just wondering)

**Comment by kerolloz** [comment]
> It will be a painful job if we decide to migrate to [Deta base](https://docs.deta.sh/docs/base/about) instead of MongoDB right? (Just wondering)

@Waelahmed99 Yup that's true. It would require a full rewrite 🥺. I also think that Deta base will not even work for our use case. Currently, we have millions of records in our submissions collection (more than 2,200,000 submissions) that we query intensively. 

Here's one of our queries:  https://github.com/icpc-scu-community/board-api/blob/3a17c938ff9a91bea95310a256702504483a0e5c/src/Builder/Submissions.builder.ts#L16-L23

I guess Render is a good option. We would only have to figure out a solution for the scrapper. But there might be another better solution. So I'm waiting for your feedback list.

**Comment by amrsalama** [comment]
@kerolloz do you have any estimation of the migration effort to **Render** and **Deta** cron? Looks like it's a matter of account creation and setup.

**Comment by kerolloz** [comment]
I guess Deta cron will not work because we connect to MongoDB from the scrapper. Deta cron is just using Deta micros to run cron jobs.
![image](https://user-images.githubusercontent.com/36763164/198849992-97026094-a4c9-46ff-9d84-7925d688b83d.png)
---
migration to Render is just a matter of account creation and setup that's true

**Comment by kerolloz** [comment]
We have another problem with Deta cron. It's that Deta micros are limited to 10 seconds of execution time. 
![image](https://user-images.githubusercontent.com/36763164/198850159-3b1684c6-0659-4a3f-a1dc-ad40311c4e0d.png)
https://docs.deta.sh/docs/micros/cron

**Comment by amrsalama** [comment]
One option (which I don't like but it's still an option) is to maintain the corn job in the same instance of the API using something like [bree](https://www.npmjs.com/package/bree), cons:

- We will need to adjust the Scrapper a little bit to work as a package and install it in the API.
- Coupling between Scrapper and API:
  - Deployment: anytime we need to update the Scrapper, we will need to publish a new version and redeploy the API.
  - Cores: if we have a single core in the API instance (which is highly likely in case of free hosting) this will affect the performance a lot!
- Multiple instances: in case we have multiple instances of the API we will have multiple Scrappers as well which might cause duplications. While it is not the case here (single instance deployment) it's still a downside we need to consider.

The bottom line is we need to find a dedicated cron service that runs NodeJS scripts 😄

**Comment by kerolloz** [comment]
@amrsalama I deployed our API to Render and it works perfectly as expected. [board-api.onrender.com](https://board-api.onrender.com)

![image](https://user-images.githubusercontent.com/36763164/198851621-715c9ef6-c629-45ab-abb4-13b83f9766ab.png)

 https://board-api.onrender.com?configs=https://api.jsonbin.io/v3/b/6175e8fc9548541c29c80745

--- 

I guess the only issue we still have is the Scrapper. But wait a moment here, Heroku didn't mention whether it's stopping Heroku Scheduler (that runs the Scrapper every 10 minutes). 

![image](https://user-images.githubusercontent.com/36763164/198851744-7529862f-e2ee-47e8-b5ee-b0283527b433.png)

We need to investigate further because if Heroku is not turning down free Scheduler then we are good to go. The API is deployed to Render and the Scrapper is working normally on Heroku Scheduler.

**Comment by kerolloz** [comment]
@Waelahmed99 any updates here?

**Comment by kerolloz** [comment]
Solved by @amrsalama in [board-scraper:server-based](https://github.com/icpc-scu-community/board-scraper/tree/server-based)
